core/local/atom_watcher.js

```diff
@@ -23,30 +31,40 @@ const log = logger({
 module.exports = class AtomWatcher {
   /*::
   syncPath: string
+  prep: Prep
+  pouch: Pouch
   events: EventEmitter
+  ignore: Ignore
   checksumer: Checksumer
+  adder: Adder
+  runner: Runner
   running: Promise<void>
   _runningResolve: ?Function
   _runningReject: ?Function
-  source: LinuxSource | WinSource
   */
 
-  constructor (syncPath /*: string */, prep /*: Prep */, pouch /*: Pouch */, events /*: EventEmitter */) {
+  constructor (syncPath /*: string */, prep /*: Prep */, pouch /*: Pouch */, events /*: EventEmitter */, ignore /*: Ignore */) {
     this.syncPath = syncPath
+    this.prep = prep
+    this.pouch = pouch
     this.events = events
+    this.ignore = ignore
     this.checksumer = checksumer.init()
 
-    // TODO do we need a debounce layer (a port of awaitWriteFinish of chokidar)?
-    const dispatcher = new Dispatcher(prep, pouch, events)
-    const checksum = new ChecksumLayer(dispatcher, this.checksumer)
+    let steps
     if (process.platform === 'linux') {
-      this.source = new LinuxSource(syncPath, checksum)
+      this.runner = this.adder = new LinuxProducer(this)
+      steps = [AddInfos, FilterIgnored, Recurse, InitialDiff, AddChecksum]
     } else if (process.platform === 'win32') {
+      this.runner = new WinProducer(this)
       // TODO add a layer to detect moves
-      this.source = new WinSource(syncPath, checksum)
+      // TODO do we need a debounce layer (a port of awaitWriteFinish of chokidar)?
+      steps = [AddInfos, FilterIgnored, InitialDiff, AddChecksum]
     } else {
       throw new Error('The experimental watcher is not available on this platform')
     }
+    let buffer = steps.reduce((buf, step) => step(buf, this), this.runner.buffer)
```

Comment:
> Could all steps be functions taking only opts and returning functions taking event batches, with no dependency on a buffer?

I don't think we can. Some steps need to examine several batches of events at the same time to do their work, like detecting a folder has moved.

> parameterize the Buffer type so the whole chain would be typechecked instead of having many optional props in the AtomWatcherEvent type

Well, I've tried and failed on the first thing: I didn't find a way to tell flow that Buffer is generic and can accept `batch` as an array of the generic type (even when always instantiated with the same type). But, we can try again later. I also think it would be really nice to type-check the events, and be sure we don't miss a mandatory field.